Analysis/AnalysisCacheProducer.py

```diff
@@ -128,20 +140,45 @@ def run_producer(
         dfw.df.Snapshot(
             f"tmp", os.path.join(workingDir, "tmp.root"), vars_to_save, snapshotOptions
         )
+        # by default cache output is saved to root file
+        # but we sometimes (BtagShapeProducer) want to save it as json
+        save_as = producer_config.get("save_as", "root")
         final_array = None
+        final_dict = None
         uproot_stepsize = producer_config.get("uproot_stepsize", "100MB")
         for array in uproot.iterate(
             f"{os.path.join(workingDir, 'tmp.root')}:tmp", step_size=uproot_stepsize
         ):  # For DNN 50MB translates to ~300_000 events
             new_array = producer.run(array)
-            if final_array is None:
-                final_array = new_array
-            else:
-                final_array = ak.concatenate([final_array, new_array])
-        check_columns(expected_columns, final_array.fields, final_array.fields)
-        with uproot.recreate(outFileName, compression=uprootCompression) as outfile:
-            outfile[treeName] = final_array
-
+            match save_as:
+                case "root":
+                    if final_array is None:
+                        final_array = new_array
+                    else:
+                        final_array = ak.concatenate([final_array, new_array])
+                case "json":
+                    # if we are saving to json file, we need to
+                    # old_dict[key] += new_dict[key] for key in keys
+                    if final_dict is None:
+                        final_dict = {key: new_array[key] for key in new_array.keys()}
+                    else:
+                        for key in final_dict.keys():
+                            final_dict[key] += new_array[key]
```

Comment:
For JSON output, you don't know how to merge A and B, because it can be producer-dependent: is it A+B or A*B? What is the nesting level to arrive at entries that are additive? etc.

The producer should provide a method that implements this logic:
```python
final_dict = producer.join(final_dict, new_array)
```